### DISCUSSION TITLE
New Plugin: Easy .htaccess

### DISCUSSION THREAD

**Discussion by gpsblues** [GetPublii/Publii#1813] [Useful Resources] [answered]
Hi everyone,

I'm excited to share a new plugin that lets you create a .htaccess file directly from Publii's interface! This is currently released as a pre-release because, to be honest, I'm no expert in Apache directives and often find myself relying on AI to avoid making a mess. 😅

[Easy Htaccess plugin](https://github.com/gpsblues/Publii-Plugin-Easy-Htaccess)

That said, I frequently encounter scenarios where I need to handle URL rewrites or explicit 404 redirects, and the manual process of creating and uploading .htaccess files gets really tedious.

In this initial version, the plugin creates an .htaccess or htaccess.txt file with essential optimizations to enhance site performance. It includes HTTPS redirection, 404 error handling, Gzip compression, and support for custom directives.

I’m sure many of you are far more knowledgeable in this area, so I’d greatly appreciate any suggestions or feedback on how to improve or fine-tune the plugin.

[Go to repository](https://github.com/gpsblues/Publii-Plugin-Easy-Htaccess)

Happy designing!

**Comment by bobmitro** [comment]
"Yo, awesome work!  Much appreciated!

**Reply by gpsblues** [reply]
Thank you so much, Bob!

By the way, I noticed that the `.htpasswd` file, even if it exists, is not transferred. Is this a limitation of Publii?

**Comment by Vinc26** [comment]
Hi there,

I've just noticed that "Redirect to HTTPS + non-www" doesn't work for here (it add's "/" and doesn't remove www well).

Finally, instead, I add manually this to get it:
```
RewriteEngine On
RewriteCond %{HTTPS} off
RewriteRule ^(.*)$ https://%{HTTP_HOST}/$1 [R,L]
RewriteCond %{HTTP_HOST} ^www\.(.+)$ [NC]
RewriteRule ^ https://%1%{REQUEST_URI} [L,R=301]
```

**Reply by gpsblues** [reply]
Hi Vinc26,

I checked the plugin using AI, and you're right, there is an error. The funny thing is that it was the AI that suggested that version to me! 😅

Your version is correct, but there could be redundancy due to two overlapping rules being applied.
I've tested this version, which seems to work fine. Could you verify it?

```
RewriteEngine On
RewriteCond %{HTTPS} off [OR]
RewriteCond %{HTTP_HOST} ^www\. [NC]
RewriteCond %{HTTP_HOST} ^(?:www\.)?(.+)$ [NC]
RewriteRule ^ https://%1%{REQUEST_URI} [L,NE,R=301]
```
I checked its validity with [htaccess.madewithlove.com](https://htaccess.madewithlove.com/), and it works correctly.

Let me know what you think!

**Reply by Vinc26** [reply]
Ok, your new code works well. :)

I haven't tested the other options. Perhaps there is also problem 🤷‍♂️

**Reply by gpsblues** [reply]
Unfortunately, I'm not very familiar with Apache and I relied a lot on AI. 
Turns out, not the best idea! :)
After your feedback, I’ve been double-checking the rules and verifying them manually.
I hope to update it soon!

**Comment by gpsblues** [comment]
Pre release v.0.0.3 available.
https://github.com/gpsblues/Publii-Plugin-Easy-Htaccess

**Comment by RSWEdev** [comment]
Using the newest version of the plugin and configure a 404 redirect works well if the unknown page is not located ina subfolder. Requesting an unknown page from a subfolder, the content from 404.html is delivered, but alls styles are lost.
I Think this is not a problem of the plugin, but of the theme/html itself: al assets are adresses relatively to the unknown subfolderlocation:
![image](https://github.com/user-attachments/assets/da20d9c5-79a8-4ab9-885f-179292e97911)

**Reply by gpsblues** [reply]
Hi, 
the plugin simply adds the `ErrorDocument 404 /404.html` directive and therefore should work. 

Unfortunately, I created the plugin with some help from the community and AI because I am inexperienced in this field. 

The redirect seems to work now. How did you solve it? Maybe I can make some changes to the plugin.

**Reply by RSWEdev** [reply]
It is not a problem of the plugin.
The asset and media folders are resolved relatively. And this does not work for subfolder levels.
I guess there is some work at template level required.

**Reply by RSWEdev** [reply, marked as answer]
Solution: just enter the host url:
![image](https://github.com/user-attachments/assets/7bd5a608-093a-4f45-8acb-0cbc19b25f51)

**Comment by gpsblues** [comment]
Hi everyone,

I’m excited to share a new feature added to the plugin! You can now manage 301 (permanent) and 302 (temporary) redirects directly from the plugin. This makes it easy to redirect old URLs to new ones without manually editing the .htaccess file.

This feature uses simple redirects (powered by mod_alias), so it's straightforward to set up. However, please note that patterns and regular expressions are not supported.

Check it out in the latest release here: [Publii Plugin Easy Htaccess vX.X](https://github.com/gpsblues/Publii-Plugin-Easy-Htaccess/releases)

As always, feel free to share feedback or report issues.

Ciao!

**Comment by nqnh** [comment]
You are great, It's very useful.
I hope the plugin can not only creates  .htaccess file but also .conf file  since there are Apache and Nginx. Is it possible?

**Reply by carstenfeller** [reply]
Nginx is very pragmatic and simple in this regard.
You create the configuration and restart Nginx to read it once.
Unlike .htaccess, the Nginx configuration isn't read every time a page is accessed, so there are no performance issues, even if you have hundreds of entries there.
Your web hosting panel should offer an editor for this, including saving and reloading.

The following lines will give you a good start without any plugin needed.

Note: There are one or two "server blocks" at the beginning of the Nginx configuration, e.g.,

`
server {
listen 80;
server_name ;
root /var/www/html/;
index index.html index.htm;
`

Don't change anything there unless you know exactly what you're doing.

**Redirects**, **browser caching**, **protect files.publii.jason**, and the **404 page** are probably important to you,
you can add these lines to the server block (the last one if there are multiple blocks).

location = /old-category/old-page/ { return 301 https://$host/new-category/new-page/; }
location = /old-category2/old-page2/ { return 301 https://$host/new-category2/new-page2/; }

location ~* \.(?:jpg|jpeg|gif|png|ico|cur|gz|svg|svgz|mp4|ogg|ogv|webm|webp|htc)$ {
expires 1d;
add_header Cache-Control "public";
} 
 
location ~* \.(?:css|js)$ { 
expires 1d; 
add_header cache control "public"; 
} 
 
location ~* \.(?:ttf|ttc|otf|eot|woff|woff2)$ { 
expires 7d; 
add_header cache control "public"; 
}

location /files.publii.json {
deny all;
}

error_page 404 /404.html;


I hope this helps.

**Reply by nqnh** [reply]
Thank you very much

**Comment by gpsblues** [comment]
Thanks so much for the feedback! Really happy to hear the plugin is helpful. 😊
About adding .conf file support: it’s tricky because those files need to be in specific server folders (not the website root), and Publii plugins write in root e in plugin subfolder.

Not sure if you’d be interested, but I recently released an update that adds cache management.

**Comment by nqnh** [comment]
Pardon,I find something is wrong Please check.
I think the right .htaccess shoud be below.
RewriteCond %{HTTPS} off [OR]
RewriteCond %{HTTP_HOST} !^www\. [NC]
RewriteCond %{HTTP_HOST} ^(?:www\.)?(.+?)$ [NC]
RewriteRule ^ https://www.%1%{REQUEST_URI} [R=301,L]

**Comment by gpsblues** [comment]
Thanks for catching this! Your solution is absolutely correct and will work perfectly.

Truth is, there are *many* ways to handle `.htaccess` redirects, and my approach (AI-assisted, I admit 😅) aims to be a **simple, compatible middle ground** for most servers.

If you prefer your version (which is just as valid), feel free to paste it into the **"Custom Directives"** section of the plugin.

You can test rules with online tools like [htaccess.madewithlove.com](https://htaccess.madewithlove.com) to verify they behave as expected.

Since the plugin is still in **beta** (a side-project of mine 🚀), I'll definitely keep your feedback in mind for future updates.

Really appreciate your input, thanks again!
